[IPsec] Enable VXLAN-in-ESP traffic

ldelossa opened the pull request:
This pull request moves the IPsec encryption hooks from pod egress to native device egress.

To accomplish this we must hairpin the traffic leaving the host back into the network stack.
When the traffic hairpins it is processed by XFRM and encryption takes place. 
Once the encryption occurs the packet is recirculated back into the stack and stack routing will push it out the proper interface. 

This achieves the goal of encrypting VXLAN traffic, in other words, the datapath will now send VXLAN-in-ESP traffic on the wire. 

```release-note
Enable VXLAN-in-ESP traffic for IPsec tunnels 
```

ldelossa committed:
ipsec,config: require devices when IPsec is enabled

To fully support Encrypted Overlay and moving IPsec encrypt hooks to the
egress device we must ensure the `to-netdev` program exists on native
devices.

Update the `DaemonConfig.AreDevicesRequired` function to return `true`
when IPsec is enabled.

This is a more general option which encompasses the `EnableIPSecEncryptedOverlay`
flag.

Moving forward Encrypted Overlay will be the default and this flag will
be removed.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
bpf,ipsec: move encryption hooks to egress device

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

ldelossa committed:
ipsec: update unit tests for VinE support

Remove the tests which are no longer valid for IPsec.

This includes the group of tests which focused on bpf_lxc and encrypted
overlay.

IPsec encryption hooks no longer run at bpf_lxc and encrypted overlay is
deprecated.

Add two new tests:
1. Integration test with bpf_host which ensures the IPsec hooks are
   called for traffic about to leave the host
2. Unit tests for the IPsec encryption hook and its hairpin
   functionality.

The latter ensures the `ctx` is prepared correctly for the hairpin and
a call to ctx_redirect is made to recirculate the packet into the stack
for XFRM processing

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	int ret = 0;
+	__u8 dst_mac = 0;
+
+	if (!eth_is_supported_ethertype(proto))
+		return DROP_UNSUPPORTED_L2;
+
```
~~We're already receiving the `proto` as parameter :).~~ this didn't make sense, I was misreading the helper

I think we should simply let such traffic pass through, no?

ldelossa replied:
Hm, this little bit of logic was ripped right from `wireguard.h`. Is it incorrect there as well?

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+
+# ifdef ENABLE_IPV6
+	case bpf_htons(ETH_P_IPV6):
+		/* ipv6 traffic not currently supported unless its tunneled */
```
huh, but don't we need to support native-routed IPv6?

ldelossa replied:
Yup, sorry this was left over during hacking. I'll fix this in a subsequent commit once I fixed all the broken testing issues :+1:

ldelossa replied:
Fixed now, along with native routing working.

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	 * so redirecting to ingress side of the same interface doesn't fail.
+	 */
+	if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
+		return DROP_INVALID;
```
You can use `THIS_INTERFACE_MAC` here.

ldelossa replied:
Ohh very nice, TIL. 

Will look into this shortly.

ldelossa replied:
No longer valid since we now redirect to `cilium_net`

julianwiedmann commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+	/* redirect to ingress side of ifindex so the packet has xfrm applied */
+	ret = ctx_redirect(ctx, ctx->ifindex, BPF_F_INGRESS);
+	if (ret != CTX_ACT_REDIRECT)
+		return DROP_INVALID;
```
I wonder if we can/should route through `to_host@cilium_host` instead, similar to what `ctx_redirect_to_proxy_host_egress()` is doing. Would that allow us to avoid the sysctl changes?

ldelossa replied:
This has probably been forgotten to history, but you were the ones to suggest just hairpining back into the egress device :).

I don't feel strongly about it. It may be nicer to only have to apply for the `accept_local` sysctl to the cilium_host interface. We'd still need the code in place to enable/disable it on the interface. Tho in this case, we'd probably care much less on the disable part...

julianwiedmann replied:
`accept_local` is already set on the `cilium_host` / `cilium_net` interfaces. See the `enableForwarding()` calls in https://github.com/cilium/cilium/blob/4c2a11690660e1e274d362b46ae95a40cf5a82ba/pkg/datapath/loader/netlink.go#L114

jschwinger233 replied:
Yeah I still remember that Joe suggested the same thing months ago when reviewing multicasting IPsec.

I didn't have a strong opinion before, but now that I understand cilium has an existing hairpin pattern (redirect from native dev to cilium_net egress), plus we don't have to set additional sysctl, so let's try.

ldelossa replied:
Yup, going to evaluate this once I get my head out of making connectivity tests work. 

My only hesitations are traversing the veth pair and not spreading this IPSec logic across more eBPF programs.
For instance, lets say we get all the way to the egress device. We can mark the IPsec packet there. 
Then, we redirect to `cilium_net` egress. We now need to look for packets marked for encryption there and let them pass thru to the ingress side of `cilium_host`, at which point we need to do the same thing, until we get to the stack. We go through a lot more code this way. 

There is the idea of just redirect to ingress side of `cilium_net`, in which case I don't think we'll traverse the veth pair as we'll look like we are coming ingress into the kernel. But this is not what's being suggested in that function pointed to.

ldelossa replied:
@julianwiedmann @jschwinger233 

I explored the simpler route of just redirect to `ingress@cilium_net`. This seems to work well and removes the requirement for the sysctl changes. 

This diff works fine:

```diff
diff --git a/bpf/lib/encrypt.h b/bpf/lib/encrypt.h
index 4fad781bb5..12b5b7e69a 100644
--- a/bpf/lib/encrypt.h
+++ b/bpf/lib/encrypt.h
@@ -233,7 +233,7 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
        struct ipv6hdr __maybe_unused *ip6;
        __u32 magic __maybe_unused = 0;
        int ret = 0;
-       __u8 dst_mac = 0;
+       union macaddr dst_mac = CILIUM_NET_MAC;

        if (!eth_is_supported_ethertype(proto))
                return DROP_UNSUPPORTED_L2;
@@ -304,20 +304,19 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto)
 #  if defined(TUNNEL_MODE) && defined(ENABLE_IPV4)
 overlay_encrypt:
 #  endif
-       /*
-        * source mac is our current egress interface, lets copy it to dmac
-        * so redirecting to ingress side of the same interface doesn't fail.
+       /* redirect to the ingress side of CILIUM_NET.
+        * this will subject the packet to the ingress XFRM hooks,
+        * encrypting the packet.
+        *
+        * the encrypted packet will be recirculated to the stack and the final
+        * egress will occur toward the IPsec tunnel's destination.
         */
-       if (eth_load_saddr(ctx, &dst_mac, 0) != 0)
-               return DROP_INVALID;
-       if (eth_store_daddr(ctx, &dst_mac, 0) != 0)
+       if (eth_store_daddr(ctx, (const __u8 *)&dst_mac, 0) != 0)
                return DROP_WRITE_ERROR;

-       /* redirect to ingress side of ifindex so the packet has xfrm applied */
-       ret = ctx_redirect(ctx, ctx->ifindex, BPF_F_INGRESS);
+       ret = ctx_redirect(ctx, HOST_IFINDEX, BPF_F_INGRESS);
        if (ret != CTX_ACT_REDIRECT)
                return DROP_INVALID;
-
        return ret;
 }
```

If the major goal here is to remove the sysctl from native devices, this will accomplish it. I'm happy to make this change. 

I don't quite see the need to traverse the VETH pair directly, as we don't gain any benefit that I can see from going thru both veth pairs.

jschwinger233 replied:
> I don't quite see the need to traverse the VETH pair directly, as we don't gain any benefit that I can see from going thru both veth pairs.

Agree, I never understood why `ctx_redirect_to_proxy_hairpin_ipv4()` redirects packets to cilium_net egress instead of cilium_host ingress directly.

ldelossa replied:
Latest push adds the change and also removes the sysctl related commit as its no longer necessary.

julianwiedmann commented on bpf/bpf_host.c:
```diff
@@ -400,11 +400,6 @@ handle_ipv6_cont(struct __ctx_buff *ctx, __u32 secctx, const bool from_host,
 	}
 
 #if defined(ENABLE_IPSEC) && !defined(TUNNEL_MODE)
-	/* See IPv4 comment. */
-	if (from_proxy && info->tunnel_endpoint && encrypt_key)
-		return set_ipsec_encrypt(ctx, encrypt_key, info->tunnel_endpoint,
-					 info->sec_identity, true, false);
-
```
This is in the packet path for `native-routing, connection was hijacked by the proxy`. Where would the encryption be applied now, in `to-netdev`?

ldelossa replied:
Does the discussion from this comment on, answer your question? https://github.com/cilium/cilium/pull/36345#issuecomment-2519545207

julianwiedmann replied:
Ok, then we need to satisfy the ordering requirement that `to-netdev` needs to be regenerated before `cilium-host` is regenerated.

Re double-encryption - I think that's a problem we need to handle gracefully from within `to-netdev` either way. There **will** be packets in-flight inside the node that are either plain ESP (native routing), or ESP-in-VXLAN (overlay routing). The new logic in `to-netdev` needs to tolerate such traffic for one release cycle.

jschwinger233 replied:
to-netdev will see:

Vxlan routing: 
- plain-text: host-to-host traffic
- plain-text-in-Vxlan: pod-to-pod traffic that requires further encryption
- ESP-in-Vxlan: old lingering encryption traffic

native routing:
- plain-text: pod-to-pod or host-to-any, that may require encryption
- ESP: old lingering traffic

julianwiedmann reviewed (commented):
took a quick pass, some comments inline

jschwinger233 commented on bpf/lib/encrypt.h:
```diff
@@ -205,6 +206,91 @@ do_decrypt(struct __ctx_buff *ctx, __u16 proto)
+		if (ctx_is_overlay(ctx)) {
+			ret = set_ipsec_encrypt(ctx, 0, ip4->daddr,
+						get_identity(ctx), false,
+						true);
```
Although I can't find a counterexample in my mind: is it a correct assumption that all pod-to-pod traffic will be encapped to vxlan in a vxlan+ipsec cilium?
Because in theory we should lookup ipcache to determine if the original packets are pod2pod, but now we go for this shortcut. Let's make sure we don't overlook any cornor case.

jschwinger233 replied:
when ipsec + egress gateway, to-netdev can see a vxlan skb but we shouldn't do ipsec encryption in this case, am I correct?

ldelossa replied:
I think you are correct since EGW utilizes VXLAN in its implementation. We may want to ask more specifically with someone who is intimate with EGW, but that seems right to me. We could be in VXLAN for EGW and I do not believe that traffic which is utilizing VXLAN for EGW purposes is subjected to encryption. We should confirm tho.

ldelossa replied:
I see that we have access to the source security ID for tunnel packets, as `bpf_overlay` sets the src ID and a FROM_OVERLAY host mark. But I'm not sure if this helps, since we'd need some idea that the traffic is for EGW purposes.

julianwiedmann replied:
`to-netdev` will see quite a bunch of VXLAN traffic that's not perfectly associated with pod-to-pod traffic (EGW, DSR-Geneve, North/South LB-to-backend, ...). WG took a pragmatic approach to this, and it resulted in [this nice list](https://docs.cilium.io/en/latest/security/network/encryption-wireguard/#which-traffic-is-encrypted) of cases :grin:.

I wouldn't give this too much thought for now - if our encryption path can transparently handle such traffic, it's probably not worth adding all sorts of special "exclude from IPsec" detection for it.

As an example for EGW: on the GW node, this traffic would *still* end up in the `from-overlay` program after decryption, and receive its EGW handling there same as before.

jschwinger233 replied:
I agree with the idea that "if our encryption path can transparently handle such traffic", but I was concerned about unexpected behavior changes: e.g. customers may assume egw won't be affected by cilium upgrade, but in fact it could: previously egw vxlan is plain-text, after upgrade it may be ipsec encrypted.

julianwiedmann replied:
Ack. I think there we are lucky that most (all?) of these "unexpected" cases are associated with `ENABLE_NODEPORT`, which is still a very **new** feature in combination with `ENABLE_IPSEC`.

ldelossa committed:
ipsec: rework encryption connectivity test for VinE traffic

IPsec VinE traffic can follow the same testing patterns as Wireguard
sans the Host-to-Host encryption parts.

Update the encryption connectivity to perform a more abstract "Encrypted
Overlay" check. This is an general term which expresses that traffic is first
tunnel encapsulated and then it is encrypted. With IPsec's move to VinE
traffic this means both WG and IPsec support encrypted overlay past
specific versions of Cilium. If the check is true then tests will sniff
on the native device and expect to see no plain tunnel packets.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>
